marshmallow_recipe/serialization.py

```diff
@@ -122,32 +91,76 @@ def load_many_v2(cls: type[_T], data: list[dict[str, Any]], *, naming_case: Nami
 
     load_many = load_many_v2
 
-    def dump_v2(
-        data: Any,
-        *,
-        naming_case: NamingCase | None = None,
-    ) -> dict[str, Any]:
-        data_schema = schema_v2(type(data), naming_case=naming_case)
+    def dump_v2(cls: type | None, data: Any, *, naming_case: NamingCase | None = None) -> dict[str, Any]:
+        data_schema = schema_v2(extract_type(data, cls), naming_case=naming_case)
         dumped, errors = data_schema.dump(data)
         if errors:
             raise m.ValidationError(errors)
         if errors := data_schema.validate(dumped):
             raise m.ValidationError(errors)
-        return dumped
+        return dumped  # type: ignore
 
-    dump = dump_v2
+    dump_impl = dump_v2
 
-    def dump_many_v2(data: list[Any], *, naming_case: NamingCase | None = None) -> list[dict[str, Any]]:
+    def dump_many_v2(
+        cls: type | None, data: list[Any], *, naming_case: NamingCase | None = None
+    ) -> list[dict[str, Any]]:
         if not data:
             return []
-        data_schema = schema_v2(type(data[0]), many=True, naming_case=naming_case)
+        data_schema = schema_v2(extract_type(data[0], cls), many=True, naming_case=naming_case)
         dumped, errors = data_schema.dump(data)
         if errors:
             raise m.ValidationError(errors)
         if errors := data_schema.validate(dumped):
             raise m.ValidationError(errors)
-        return dumped
+        return dumped  # type: ignore
 
-    dump_many = dump_many_v2
+    dump_many_impl = dump_many_v2
 
 EmptySchema = m.Schema
+
+
+@overload
+def dump(
+    data: Any,
+    *,
+    naming_case: NamingCase | None = None,
+) -> dict[str, Any]: ...
+
+
+@overload
+def dump(
+    cls: type,
+    data: Any,
+    *,
+    naming_case: NamingCase | None = None,
+) -> dict[str, Any]: ...
+
+
+def dump(*args: Any, **kwargs: Any) -> dict[str, Any]:
+    if len(args) == 1:
+        return dump_impl(None, *args, **kwargs)
```

Comment:
call with `data=` will not work

```
dump(value)  # works
dump(data=value)  # TypeError: dump_v3() missing 1 required positional argument: 'cls'
```

```
dump(Value, value)  # works
dump(cls=Value, data=value)  # works
dump(Value, data=value)  # TypeError: dump_v3() got multiple values for argument 'data'
```

we can add `/` or we can check cls and data in args and kwargs and check if argument is type or dataclass

Reply:
https://github.com/anna-money/marshmallow-recipe/pull/172